### DISCUSSION TITLE
Should a state live in a procedure environment?

### DISCUSSION THREAD

**Discussion by ceving** [vanjs-org/van#231] [General]
I am still struggling with web development and I would like to discuss the following observation.

Imagine a web application, which is a sequence of forms each containing a sequence of input fields. The DOM might look like this:

- App
  1. Form
      1. Input
      2. Input
  2. Form
      1. Input
      2. Input

Every input should be validated, every form should be validated and all forms should be validated. Because the input has an impact on the validation, both are states. When I put each state in a variable, which lives in an environment of a procedure, each validity can only be defined after the corresponding input, because the validity depends on the input, the same for the combined validities. In the end the definitions will have the following order:

~~~
var app_form1_input1
var app_form1_input1_validity
var app_form1_input1
var app_form1_input2_validity
var app_form1_validity
var app_form2_input1
var app_form2_input1_validity
var app_form2_input1
var app_form2_input2_validity
var app_form2_validity
var app_validity
~~~

This is exactly the opposite order of the DOM tree.

VanJS advocates two aspects:

1. Procedural composition of the DOM tree
2. Storing states in procedure environments

But the shape of the procedures building the DOM tree is contrary to the shape of the procedures working on the states. How does this fit together?

The alternative would be to store the states right in the DOM tree. Every element, which needs validation, can store the validation state in its dataset. And when a validity of an input element has changed the element can emit a bubbling custom event. It will bubble the DOM tree up and every element on its way can react to the event by updating its validation state. This works right out of the box without any help of a toolkit. Is seems to me that for this particular use-case a DIY solution is much easier than a solution using VanJS.

**Comment by Atmos4** [comment]
It feels like your thinking is a bit more complicated than it needs to be.

> Every input should be validated

This is not true. You will give yourself a lot more trouble if you try to validate all inputs all the time. In the end when it comes to data submission, only backend validation matters. Frontend validation is only something nice for the user that you can perform sporadically.

> procedures

I don't understand your usage of the word `procedure`. If you want granular input validation, you are perfectly fine doing it in the input field level with `van.state` and care about global form validation on submit.

> all forms

Given the way HTML forms work, thinking about form submission with multiple forms on the same page doesn't make a lot of sense, since only one form should be submitted at the time. If you have many forms you want to validate and submit at the same time, it might be that you are thinking about form submission in a wrong way.

Keep in mind doing any kind of validation in the frontend (i.e. Javascript) is not secure as anyone could bypass it. It should just be viewed as a nice UX bonus, nothing more.

PS: You might want to take a look at [this community add-on](https://github.com/vanjs-org/van/discussions/219) dedicated to form state management with VanJS.

**Reply by ceving** [reply]
For me front-end validation is mandatory. You are right, that it does not provide any security. But it reduces the load of the servers. I do not want to waste server resources for stupid inputs. I do not have to pay for the computing time on the client, but I have to pay for the computing time on the server. The following does full input validation (1 app, 2 forms, 7 input fields) without any toolkit. And it keeps all states in the DOM.
~~~html
<!DOCTYPE html>
<html>
  <head>
    <title>Event</title>
    <style>
      body {
        background-color: floralwhite;
      }
      main, form, input[type="text"] {
        border: 2px solid silver;
        margin: 1ex;
      }
      main[data-validity="true"],
      form[data-validity="true"],
      input[data-validity="true"] {
        border: 2px solid green;
      }
      main[data-validity="false"],
      form[data-validity="false"],
      input[data-validity="false"] {
        border: 2px solid red;
      }
    </style>
  </head>
  <body>
    <main id="main">
      <form id="f1">
        <input type="checkbox" id="f1i1"/><label>store</label>
        <input type="checkbox" id="f1i2"/><label>share</label>
        <input type="checkbox" id="f1i3"/><label>unbid</label>
      </form>
      <form id="f2">
        <input type="text" id="f2i1" placeholder="3 digits"/>
        <input type="text" id="f2i2" placeholder="2 digits >=18"/>
        <input type="text" id="f2i3" placeholder="2 digits 1-52"/>
        <input type="text" id="f2i4" placeholder="7 digits"/>
      </form>
    </main>
    <script type="module">
      function nil(arg) {
        const b = arg === undefined || arg === null || arg === ''; }
      function digits(n) {
        const q = n === undefined ? '+' : `{${n}}`;
        const rx = new RegExp(`^\\d${q}\$`);
        return str => {
          if (nil(str)) return undefined;
          return rx.test(str); }}
      function gtoe(n) {
        return str => {
          if (nil(str)) return undefined;
          const x = parseInt(str);
          if (x === NaN) return undefined;
          return x >= n; }}
      function range(n, m) {
        return str => {
          if (nil(str)) return undefined;
          var x = parseInt(str);
          if (x === NaN) return undefined;
          return x >= n && x <= m; }}
      function and(c1, c2) {
        return str => {
          const b1 = c1(str);
          if (nil(b1)) return undefined;
          if (b1 === false) return false;
          const b2 = c2(str);
          if (nil(b2)) return undefined;
          return b1 && b2; }}
      function checkbox_validation(target, check) {
        if (typeof target === 'string')
          target = document.getElementById(target);
        target.addEventListener('input', (ev) => {
          const validity = `${check(target.checked)}`;
          if (validity != target.dataset.validity) {
            target.dataset.validity = validity;
            const cev = new CustomEvent("validation", { bubbles: true });
            target.dispatchEvent(cev); }}); }
      function text_validation(target, check) {
        if (typeof target === 'string')
          target = document.getElementById(target);
        target.addEventListener('input', (ev) => {
          const validity = `${check(target.value)}`;
          if (validity != target.dataset.validity) {
            target.dataset.validity = validity;
            const cev = new CustomEvent("validation", { bubbles: true });
            target.dispatchEvent(cev); }}); }
      function validation(target, check) {
        if (typeof target === 'string')
          target = document.getElementById(target);
        target.addEventListener("validation", (ev) => {
          const validity = `${check()}`;
          if (validity != target.dataset.validity)
            target.dataset.validity = validity; }); }
      function validity(element) {
        if (typeof element === 'string')
          element = document.getElementById(element);
        switch (element.dataset.validity) {
        case 'true': return true;
        case 'false': return false;
        default: return undefined; }}
      validation('main', () => validity('f1') && validity('f2'));
      validation('f1', () =>
        validity('f1i1') && validity('f1i2') && validity('f1i3'));
      checkbox_validation('f1i1', c => c === true);
      checkbox_validation('f1i2', c => c === true);
      checkbox_validation('f1i3', c => c === true);
      validation('f2', () =>
        validity('f2i1') && validity('f2i2') && validity('f2i3') && validity('f2i4'));
      text_validation('f2i1', digits(3));
      text_validation('f2i2', and(digits(2), gtoe(18)));
      text_validation('f2i3', and(digits(2), range(1,52)));
      text_validation('f2i4', digits(7));
    </script>
  </body>
</html>
~~~
And I have some trouble to achieve the same with the "help" of VanJS. I find it in particular difficult to move the states from the DOM into variables, because of the contrary organization. I can not write something like this:
~~~javascript
main({id:"main"},
     () => {
       var main_validity = van.derive(() => f1_validity.val && f2_validity.val);
     },
     form({id:"f1"},
          input({id:"f1i1"}),
          input({id:"f1i2"}),
          input({id:"f1i3"})),
     form({id:"f2"},
          input({id:"f2i1"}),
          input({id:"f2i2"}),
          input({id:"f2i3"}),
          input({id:"f2i4"})));
~~~
Because `main_validity` can not see `f1_validity` or any other dependency, because the dependencies are coming after the state, which depends on them. This problem occurs only, if you put states into procedure scopes. The DOM solution does not have this problem, because all of the DOM is global, there does not exist any scope. The problem does not get easier by the introduction of scopes, it gets more complicated.

**Reply by Tao-VanJS** [reply]
> This problem occurs only, if you put states into procedure scopes.

You can definitely put all the states into the global scope. I don't see a problem with global-scoped states.

**Reply by ceving** [reply]
Yes but then I have the same situation as in the past. I have the definition of the DOM in one place and all the logic in another place. I would like to have some kind of encapsulation. The creation of the DOM elements and the logic, which deals with the elements and the events they produce, belong together. The essence of the whole problem is, that the creation of DOM elements is a top-down job, but managing state-changes is a bottom-up job. When I put states into the DOM, while the DOM gets created in order to read the states with `getElementByID` this works, because `getElementById` implements some kind of lazy evaluation. But when I put the state into the environment of a procedure, I give up the laziness, because JavaScript is eager. I think the makes the whole idea of having states in ordinary JavaScript variables questionable.

**Reply by ceving** [reply]
BTW: when I say procedure I mean something which behaves in this way:
https://sarabander.github.io/sicp/html/3_002e2.xhtml#g_t3_002e2

**Reply by ceving** [reply]
I think I would like to be able to write something like this.
~~~javascript
main({id:'main'},
     state({id:'validity'},
           () => state_set('validity', and(state_get('f1', 'validity'),
                                           state_get('f2', 'validity')))),
     form({id:'f1'},
          state({id:'validity'},
                () => state_set('validity', and(state_get('i1', 'validity'),
                                                state_get('i2', 'validity'),
                                                state_get('i3', 'validity')))),
          input({id:'i1', oninput: e => state_set('validity', validate_f1i1(e.target.value))},
                state({id:'validity'})),
          input({id:'i2', oninput: e => state_set('validity', validate_f1i2(e.target.value))},
                state({id:'validity'})),
          input({id:'i3', oninput: e => state_set('validity', validate_f1i2(e.target.value))},
                state({id:'validity'}))),
     form({id:'f2'},
          state({id:'validity'},
                () => state_set('validity', and(state_get('i1', 'validity'),
                                                state_get('i2', 'validity'),
                                                state_get('i3', 'validity'),
                                                state_get('i4', 'validity')))),
          input({id:'i1', oninput: e => state_set('validity', validate_f2i1(e.target.value))},
                state({id:'validity'})),
          input({id:'i2', oninput: e => state_set('validity', validate_f2i2(e.target.value))},
                state({id:'validity'})),
          input({id:'i3', oninput: e => state_set('validity', validate_f2i3(e.target.value))},
                state({id:'validity'})),
          input({id:'i4', oninput: e => state_set('validity', validate_f2i4(e.target.value))},
                state({id:'validity'}))));
~~~
I it verbose to make it understandable.

But in general the variable lookup needs to be done in the opposite direction as it is usually done when procedures are evaluated. Procedures search for missing variables in their parent environment. But in the DOM a parent needs to lookup variables in its children.

**Reply by Tao-VanJS** [reply]
I am not sure if I can express my feeling about this. It seems that you're trying to pursuit the perfect solution based on your specific criteria for a very special-case scenario. But IMHO, day-to-day programming primarily focuses on getting the problem solved in a reasonably maintainable way. Sometimes there are trade-offs but programmers just move on.

**Comment by Tao-VanJS** [comment]
> But the shape of the procedures building the DOM tree is contrary to the shape of the procedures working on the states. How does this fit together?

Not sure what you meant by this. Do you have a concrete example where things are running into troubles?
